### DISCUSSION TITLE
Equivalent of "dig +search" flag with NAT64

### DISCUSSION THREAD

**Discussion by tom-brown-kr** [openthread/openthread#9345] [Q&A] [answered]
I am using the border router's upstream DNS and `$ sudo ot-ctl nat64 state` shows that it is active. My Thread child device can resolve hosts like google.com without issue. The upstream DNS uses "iot-edge" as an alias for "pc7.ab.cd.com" but if my child device does a DNS lookup for "iot-edge" using `otDnsClientResolveIp4Address()` it gets a "No such name error" as seen below.
![image](https://github.com/openthread/openthread/assets/138053999/4fd356fe-1462-4662-9b02-7b0b994d31ba)

Similarly, I get an NXDOMAIN error on my Raspberry Pi border router when I try `$ dig iot-edge`
![image](https://github.com/openthread/openthread/assets/138053999/cfb27f1d-8055-4377-aa1f-00cee64033cf)

However, the Pi is able to resolve "iot-edge" when I add the `+search` flag.
![image](https://github.com/openthread/openthread/assets/138053999/2f1ae787-21af-4d05-a946-4c84eefb0d21)

**What setting am I missing in my border router's NAT64 setup to duplicate the `+search` flag's functionality and enable these alias lookups?** Thank you in advance!

**Comment by tom-brown-kr** [comment]
It seems that the `+search` flag enables use of the domain directive in `/etc/resolv.conf`. I see `otSrpClientSetDomainName()` - maybe this is what I should be using? However, it would be much easier if I could set the domain name on the border router side because the Thread child has no knowledge of the border router's `/etc/resolv.conf` domain directive
![image](https://github.com/openthread/openthread/assets/138053999/bb2ed6b4-56fa-4cc2-b5aa-4a9bfe614827)

**Reply by abtink** [reply]
`otSrpClientSetDomainName()` sets the domain name to be used by SRP Client on the device. It won't impact or change DNS client code or behavior.

**Reply by tom-brown-kr** [reply]
Thank you, good to know. Maybe `$ sudo ot-ctl srp server domain [domain-name]` is what I am looking for on the border router? I can `grep` the domain from `/etc/resolv.conf`

**Reply by tom-brown-kr** [reply]
I also tried using the CLI command `srp server domain [domain-name]` but that did not solve my issue. If my Thread child device does a DNS lookup for "example" it gets the following response:

"Standard query response 0xabcd No such name A example SOA a.root-servers.net"

However, `$ nslookup example` and `$ dig +search example` on the border router yield the desired IPv4 address. Is there anything else I should be trying or investigating?

**Reply by abtink** [reply]
I am not familiar with this but I guess the idea is to replace the zone/domain name with another one?
i.e.,  instead of querying for `{name}.example.com` the query should be for `{name}.{new-zone/domain}`. 
You can try determining the new name manually and try calling API with this new name.

**Reply by abtink** [reply]
BTW, I know SOA records are not parsed by DNS client implementaion so whatever feature/behavior is using this is not supported.

**Reply by tom-brown-kr** [reply]
I think it might help if I re-frame my original question.

I have a host with an alias "iot-edge" and domain "my.domain.com." The border router can determine this domain name (its /etc/resolv.conf file is auto-generated with `domain my.domain.com` in the first line of the file). However, the Thread child has no knowledge of this domain name.

If the Thread child does a DNS lookup for "iot-edge.my.domain.com" then it successfully receives this host's IPv4 address from the border router's upstream DNS server. However, if the Thread child does a DNS lookup for simply "iot-edge" then it receives the "No such name" DNS lookup error.

By setting `> srp server domain my.domain.com` I was hoping that the border router would automatically append this domain to DNS requests (or handle it in another way that is functionally equivalent). So, when the Thread child does a DNS lookup for "iot-edge" the border router appends "my.domain.com" in the message to its upstream DNS server and an IPv4 address is successfully received. Or, the Thread child could somehow do a lookup of the border router's srp server domain name setting and do this appending itself.

**Reply by abtink** [reply, marked as answer]
`srp server domain` is intended to set the domain name used by SRP service registration for *the devices within the Thread mesh*. It does not apply to DNS client queries and change behavior of DNS server/resolver.

----

Some thoughts/suggestions:

- You can check the OT DNS-SD server code `dnssd_serevr.h/hpp/cpp` and its feature `OPENTHREAD_CONFIG_DNS_UPSTREAM_QUERY_ENABLE` which allows recursive resolving (passing on the query to another entity to be resolved using platform `otPlatDnsStartUpstreamQuery()` APIs). One possibility I see is that you can update the platform implementation of this API to try to update the query message and append the domain name (i.e., check if query name in the query message does not have domain, and if so replace it with domains it knows should be used). Though I am not sure if this is a good solution in general since I expect/assume that the full name should be specified by client in the constructed DNS query.

- To me, the better/proper solution is to implement a mechanism for Thread child (device) to learn the list of domains it should search and have it append the domain names to the host when it tries to perform query.

**Reply by tom-brown-kr** [reply]
Thank you for the input, I was starting to suspect that I was using `srp server domain` incorrectly. I will pursue option 2 that you have listed. Appreciate the help!
